0000-stream/0000-stream.md

```diff
@@ -0,0 +1,182 @@
+
+Multiple Money Streams MAY be used to separately account for value sent and received for different purposes. A Money Stream can be long-lived and the send and receive limits MAY be adjusted over time (for example, in response to an application-level action) or a Money Stream can be used to account for a single logical payment, in which case it would be closed when the payment is finished.
+
+Money Streams are identified by integers starting from 1. Odd-numbered Money Streams are initiated by the Client and even-numbered Money Streams are initiated by the Server. Money Stream identifiers are unrelated to Data Stream identifiers and a single Connection MAY have Money Streams and Data Streams with the same number.
+
+### Multiplexed Data Streams
+
+STREAM also offers the ability to send data through the Connection over ILP (rather than over the Internet). Data Streams are a lightweight byte-stream abstraction used for grouping data sent over the connection. They provide reliable, in-order delivery of bytes and may divide messages amongst multiple ILP packets. Data Streams are heavily inspired by [QUIC Bidirectional Streams](https://tools.ietf.org/html/draft-ietf-quic-transport-10#section-10).
+
+Data Streams may be long-lived or they may be used for individual requests and/or response messages. For details on this, see [Streams: QUIC's Data Structuring Abstraction](https://tools.ietf.org/html/draft-ietf-quic-transport-10#section-10).
+
+Data Streams are identified by integers starting from 1. Odd-numbered Data Streams are initiated by the Client and even-numbered Data Streams are initiated by the Server. Data Stream identifiers are unrelated to Money Stream identifiers and a single Connection MAY have Data Streams and Money Streams with the same number.
+
+### Stream and Connection Flow Control
+
+STREAM implements stream- and connection-level flow control. Clients and Servers communicate the maximum amount of money or data they are willing to receive on a given stream or for the connection as a whole. Limits may be changed and retransmitted throughout the life of the Connection to accept more money or data as desired by the application.
+
+Money Streams SHOULD start with send and receive limits of zero to prevent accidental sending.
+
+Data Streams MAY start with non-zero limits but should set the limits such that out-of-order incoming data will not exceed the buffer capacity allocated.
+
+Connections SHOULD limit the number of Money and Data Streams to prevent a malicious Sender from opening an excessive number of streams to exhaust the Receiver's memory.
+
+### Connection Migration
+
+Clients and Servers MAY change the address the other entity will send packets to at any time by sending a `ConnectionNewAddress` frame with their new ILP address.
+
+If one entity changes their address, the other SHOULD NOT assume that the asset is the same or that the exchange rate will be similar.
+
```

Comment:
Given that STREAM Connections can be long-lived, did you consider the implications for the encryption key? A given key should only be used to encrypt a limited number of messages. 

That number is large, but it might become relevant for long-lived connections. Quote from the NIST publication on AES-GCM:
> The  total  number  of  invocations  of  the  authenticated  encryption  function  shall  not  exceed 2^32, including all IV lengths and all instances of the authenticated encryption function with the given key.

Reply:
Good point. Do you think it's better to drop the connection or add a frame that allows you to change the shared secret? I originally thought about adding a shared secret to the `ConnectionNewAddress` frame but opted against it.

Reply:
Let's break down the pros and cons of both options. 

The options are: After the total number of encrypted messages exceeds some threshold, 1) the connection should either be dropped or 2) the encryption key needs to be renewed.

**1) Dropping the connection**
\+ simple to implement
\- higher-level protocols have to assume that the connection might be dropped and have to reestablish

**2) Renewing the key**
\+ no connection interruption due to exhausting the key/nonces
*\- More complex.* Requires a new frame type or an existing frame needs to be extended to communicate the new key
*\- Duplicates Logic.* Key establishment is explicitly not part of PSK but assumed to be done over a separate, secure communication channel. If we add a way to exchange keys within the protocol, we are essentially duplicating the key establishing logic.
*\- Potentially Insecure.* Depending on how it is implemented, it might be insecure. For example, a man-in-the-middle could just count how many messages went over a connection and when it is about time to renew the key, he would drop all packets, thus preventing client and server to exchange a new key. Once the max. number of invocations for AES-GCM is reached, client/server have no other (secure) option as to drop the connection. Otherwise, the man-in-the-middle can decrypt packets and learn the new encryption key.

The last point makes me realize that as a last resort the connection NEEDS to be dropped. Renewing the key within the protocol is just a convenience feature. So I definitely recommend to go with 1) for now.